services/wallet/async/async_test.go

```diff
@@ -0,0 +1,59 @@
+package async
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAtomicGroupTerminatesOnOneCommandFailed(t *testing.T) {
+	ctx := context.Background()
+	group := NewAtomicGroup(ctx)
+
+	err := errors.New("error")
+	group.Add(func(ctx context.Context) error {
+		time.Sleep(1 * time.Millisecond)
+		return err // failure
+	})
+	group.Add(func(ctx context.Context) error {
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-time.After(1 * time.Millisecond): // wait for the first command to fail and terminate the second command
+				_ = 1
+			}
+		}
+	})
+
+	<-group.WaitAsync()
+	require.Equal(t, err, group.Error())
+}
+
+func TestAtomicGroupWaitsAllToFinish(t *testing.T) {
+	ctx := context.Background()
+	group := NewAtomicGroup(ctx)
+
+	finished := false
+	group.Add(func(ctx context.Context) error {
+		time.Sleep(1 * time.Millisecond)
```

Comment:
Same as above, how about replacing this sleep with a different order of adding the commands? Because we know the other command needs 5 ms, if this is added after the other we know it has a chance to run without a sleep.

Reply:
The comment was based on reading the old description :). Thanks for clarifying.